### DISCUSSION TITLE
Federation status of public posts may leak information about protected/private replies

### DISCUSSION THREAD

**Discussion by nemobis** [mastodon/mastodon#28536] [Q&A]
### Pitch

My instance will helpfully pre-fetch the parent posts of any posts I receive. This is nice to more easily follow threaded conversations, but confusing when part of the conversation is public and part is private. In particular, a "protected" reply to a "public" post shouldn't cause the public post to be shown on the federated timeline, because it will then be revealed in public that _someone_ has privately interacted with that post.

### Motivation

Users of small instances and the people they interact with are more vulnerable to a (sort of) de-anonymization attack, at least with default Mastodon instance settings. On a single-user instance, if I (as a visitor) see a post in the federated timeline (e.g. a tag search), I'll probably guess that the instance's user has interacted with that post; but if I'm the author, and I know I have no followers on that instance, I'll know that someone else must have forward the post there.

This problem is in a way the opposite of https://github.com/mastodon/mastodon/issues/24193 / https://github.com/mastodon/mastodon/issues/19685 / https://github.com/mastodon/mastodon/issues/24523. It may be considered expected behaviour, or not. It might also be alleviated on the client side by using different visibility defaults or by sending warnings of what happens if you reply to a public post with a private/protected post or vice versa.

**Comment by nemobis** [comment]
https://github.com/mastodon/mastodon/issues/23406 is a bit of the inverse issue (also has several links to related issues). This issue is different because it's not about who has the permission to see a post (the original post is public).

**Comment by trwnh** [comment]
> In particular, a "protected" reply to a "public" post shouldn't cause the public post to be shown on the federated timeline, because it will then be revealed in public that someone has privately interacted with that post.

Not necessarily? That public post may have ended up on your server because someone searched for it directly, without interacting with it in any other way.

**Comment by nemobis** [comment]
> That public post may have ended up on your server because someone searched for it directly

It used to be the case that any unregistered user could cause a post to be federated on your instance by searching it through its URL, but that's no longer true because of the other issue I reported, https://github.com/mastodon/mastodon/issues/28376 (which may be either a bug or a feature). So the presence of the post reveals that a local user was aware of it (and did _something_ with it, whether or not it was a recorded action like reply, start or boost), doesn't it?

**Comment by trwnh** [comment]
> So the presence of the post reveals that a local user was aware of it (and did something with it, whether or not it was a recorded action like reply, start or boost), doesn't it?

Still nope. It could be as indirect as a local user following a remote account which boosted it. Or it could even be that a local user follows a remote account which made a post that mentioned a different account that hasn't been seen before and has pinned some public posts. There's no indication or implication that a local user did anything. That's not really leaking any information. Even in cases you describe above, the "leak" is incredibly minor IMO.
